File: core/captcha/src/main/java/com/anji/captcha/service/impl/CaptchaServiceFactory.java
package com.anji.captcha.service.impl;

import com.anji.captcha.service.CaptchaCacheService;
import com.anji.captcha.service.CaptchaService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.HashMap;
import java.util.Map;
import java.util.Properties;
import java.util.ServiceLoader;

/**
 * Created by raodeming on 2020/5/26.
 */
public class CaptchaServiceFactory {

    private static Logger logger = LoggerFactory.getLogger(CaptchaServiceFactory.class);

    public static CaptchaService getInstance(Properties config){
        return instances.get(config.getProperty("captcha.type", "default"));
    }

    public static CaptchaCacheService getCache(String cacheType){
        return cacheService.get(cacheType);
    }

    public volatile static Map<String,CaptchaService> instances = new HashMap();
    public volatile static Map<String,CaptchaCacheService> cacheService = new HashMap();
    static {
        ServiceLoader<CaptchaCacheService> cacheServices = ServiceLoader.load(CaptchaCacheService.class);
        for(CaptchaCacheService item : cacheServices){
                cacheService.put(item.type(), item);
        }
        logger.info("supported-captchaCache-service:{}", cacheService.keySet().toString());
        ServiceLoader<CaptchaService> services = ServiceLoader.load(CaptchaService.class);
        for(CaptchaService item : services){
            instances.put(item.captchaType(), item);
        };
        logger.info("supported-captchaTypes-service:{}", instances.keySet().toString());
    }
}

File: service/golang/src/service/CaptchaServiceFactory.go
package service

import (
	constant "anjiplus/captcha/const"
	"anjiplus/captcha/model"
	"fmt"
)

type CaptchaServiceFactory struct {
	cacheService   map[string]*CaptchaCacheService
	captChaService map[string]*CaptchaService
}

func (f *CaptchaServiceFactory) GetCache(cache string) CaptchaCacheService {
	if f.cacheService[cache] != nil {
		return *f.cacheService[cache]
	}
	panic(model.ResponseModel{RepCode: constant.CAPTCHA_CACHETYPE})
}
func (f *CaptchaServiceFactory) SetCacheService(s CaptchaCacheService) {
	f.cacheService[s.CacheType()] = &s
}

func (f *CaptchaServiceFactory) SetCaptchaService(s CaptchaService) {
	f.captChaService[s.CaptchaType()] = &s
}
func init() {
	fmt.Println("captcha-service-factory-init....")
}
